src/QirRuntime/public/ChunkedStore.hpp

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <assert.h>
+#include <vector>
+
+#include "BitStates.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    /*==============================================================================
+        Provides dynamically extendable storage for TItem_WithId (usually, qubits)
+        that never moves previously allocated items, meaning the address of the item
+        can be safely used to access it until it's released.
+
+        The backing memory is split into contiguous chunks of fixed size, but the
+        chunks aren't guaranteed to be located next to each other.
+
+        TItem_WithId must provide public field "id". The store uses the id as index
+        to speedup look up of items.
+
+        TODO:
+        1. is it beneficial to release empty chunks?
+        2. do we really need to store the id on the items?
+        3. should we shrink chunks vector?
+    ==============================================================================*/
+    template <int COUNT_ITEMS, typename TItem> struct Chunk
+    {
+        static_assert(
+            COUNT_ITEMS < sizeof(uint64_t) * 8 + 1,
+            L"keeping track of filled should be reviewed if this assert fails");
+
+        uint64_t filled = 0;
+        TItem data[COUNT_ITEMS];
+
+        bool IsValidItem(long index)
+        {
+            return (index < COUNT_ITEMS) && (filled & (static_cast<uint64_t>(1) << index));
+        }
+    };
+
+    template <typename TItem_WithId> struct CChunkedStore

```

Comment:
Why not just use `vector<TItem_WithId>`? Yes, it will be slower when it needs to reallocate, but as far as I can tell this class is only used for testing, so the perf hit isn't particularly meaningful, and this class may be slightly slower when accessing elements, so it may all balance out anyway?